flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerService.java

```diff
@@ -18,43 +18,306 @@
 package org.apache.flink.streaming.api.operators;
 
 import org.apache.flink.annotation.Internal;
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.core.memory.DataInputViewStreamWrapper;
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+import org.apache.flink.runtime.state.KeyGroupRange;
+import org.apache.flink.streaming.runtime.tasks.EventTimeCallback;
+import org.apache.flink.streaming.runtime.tasks.ProcessingTimeCallback;
+import org.apache.flink.streaming.runtime.tasks.ProcessingTimeService;
+import org.apache.flink.util.InstantiationUtil;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ScheduledFuture;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
 
 /**
  * Interface for working with time and timers.
  *
  * <p>This is the internal version of {@link org.apache.flink.streaming.api.TimerService}
  * that allows to specify a key and a namespace to which timers should be scoped.
  *
+ * All d
+ * 
+ * @param <K> Type of the keys in the stream
  * @param <N> Type of the namespace to which timers are scoped.
  */
 @Internal
-public interface InternalTimerService<N> {
+public abstract class InternalTimerService<K, N> implements ProcessingTimeCallback, EventTimeCallback {
```

Comment:
I would suggest to still keep the old interface and rename this to `AbstractInternalTimerService<K, N> implements InternalTimerService<N>`. Like that, we don't need to introduce the generic parameter K in all places, which is actually giving away some implementation detail (K is used only for a member, not for the interface methods). I also like to keep the interface slim, and probably not every code that deals with `InternalTimerService` has to see all the methods, e.g. for snapshots.